sw/device/lib/runtime/print.c

```diff
@@ -68,6 +71,151 @@ void base_set_stdout(buffer_sink_t out) {
   base_stdout = out;
 }
 
+static const size_t kSpiDeviceReadBufferSizeBytes =
+    SPI_DEVICE_PARAM_SRAM_READ_BUFFER_DEPTH * sizeof(uint32_t);
+static const size_t kSpiDeviceFrameHeaderSizeBytes = 12;
+static const size_t kSpiDeviceMaxFramePayloadSizeBytes =
+    kSpiDeviceReadBufferSizeBytes - kSpiDeviceFrameHeaderSizeBytes - 4;
+static const uint32_t kSpiDeviceFrameMagicNumber = 0xa5a5beef;
+static uint32_t spi_device_frame_num = 0;
+
+/**
+ * Sends data out of the SPI device.
+ *
+ * Data is packaged into a frame that is described below.
+ * The host side reads the header first, then decides how many words
+ * to read from the data section.
+ *
+ * -----------------------------------------------
+ * |      Magic Number     | 4-bytes  |          |
+ * -----------------------------------|          |
+ * |      Frame Number     | 4-bytes  |  Header  |
+ * -----------------------------------|          |
+ * |   Data Length (bytes) | 4-bytes  |          |
+ * -----------------------------------|----------|
+ * |      Data (word aligned)         |          |
+ * -----------------------------------|   Data   |
+ * |     0xFF Pad Bytes    | <4-bytes |          |
+ * -----------------------------------|----------|
+ */
+static size_t spi_device_send_frame(void *data, const char *buf, size_t len) {
+  dif_spi_device_handle_t *spi_device = (dif_spi_device_handle_t *)data;
+  const size_t kDataPacketSizeBytes = ((len + 3u) & ~3u);
+  const size_t kFrameSizeBytes =
+      kSpiDeviceFrameHeaderSizeBytes + kDataPacketSizeBytes;
+  uint8_t frame_bytes[kFrameSizeBytes];
```

Comment:
This is a potentially large buffer on the stack? Is it possible to directly write into the SPI buffer, at least for the payload? (the header could be constructed on the stack)

Reply:
Yes, a ~2KB buffer will be allocated on the stack in the worst case. To optimize memory usage, the implementation has been modified to allow for the direct writing of aligned data into the SPI buffer.  A frame is split into three parts:
1. Header: A 12-byte array is constructed on the stack
2. Aligned Data: This data is written directly into the SPI buffer
3. Unaligned Data: A 4-byte array is constructed on the stack to hold this data.